stan/math/prim/mat/prob/ordered_logistic_glm_lpmf.hpp

```diff
@@ -0,0 +1,212 @@
+                          N_instances);
+  }
+  check_consistent_size(function, "Weight vector", beta, N_attributes);
+  check_bounded(function, "Vector of dependent variables", y, 1, N_classes);
+  check_ordered(function, "Cut-points", cuts);
+  check_finite(function, "Final cut-point", cuts[N_classes - 2]);
+  check_finite(function, "First cut-point", cuts[0]);
+
+  if (size_zero(y, cuts))
+    return 0;
+
+  if (!include_summand<propto, T_x_scalar, T_beta_scalar, T_cuts_scalar>::value)
+    return 0;
+
+  T_partials_return logp(0);
+  const auto& x_val = value_of_rec(x);
+  const auto& beta_val = value_of_rec(beta);
+  const auto& cuts_val = value_of_rec(cuts);
+
+  const auto& beta_val_vec = as_column_vector_or_scalar(beta_val);
+  const auto& cuts_val_vec = as_column_vector_or_scalar(cuts_val);
+
+  scalar_seq_view<T_y> y_seq(y);
+  Array<double, Dynamic, 1> cuts_y1(N_instances), cuts_y2(N_instances);
+  for (int i = 0; i < N_instances; i++) {
+    int c = y_seq[i];
+    if (c != N_classes) {
+      cuts_y1[i] = cuts_val_vec[c - 1];
+    } else {
+      cuts_y1[i] = INFINITY;
+    }
+    if (c != 1) {
+      cuts_y2[i] = cuts_val_vec[c - 2];
+    } else {
+      cuts_y2[i] = -INFINITY;
+    }
+  }
+
+  T_location location = x_val * beta_val_vec;
+  if (!isfinite(sum(location))) {
+    check_finite(function, "Weight vector", beta);
+    check_finite(function, "Matrix of independent variables", x);
+  }
+
+  Array<double, Dynamic, 1> cut2 = location - cuts_y2;
+  Array<double, Dynamic, 1> cut1 = location - cuts_y1;
+
+  // Not immediately evaluating next two expressions benefits performance
+  auto m_log_1p_exp_cut1
+      = (cut1 > 0.0).select(-cut1, 0) - (-cut1.abs()).exp().log1p();
+  auto m_log_1p_exp_m_cut2
+      = (cut2 <= 0.0).select(cut2, 0) - (-cut2.abs()).exp().log1p();
+
+  if (is_vector<T_y>::value) {
+    Eigen::Map<const Eigen::Matrix<int, Eigen::Dynamic, 1>> y_vec(&y_seq[0],
+                                                                  y_seq.size());
+    logp = y_vec.cwiseEqual(1)
+               .select(m_log_1p_exp_cut1,
+                       y_vec.cwiseEqual(N_classes).select(
+                           m_log_1p_exp_m_cut2,
+                           m_log_1p_exp_m_cut2 + log1m_exp(cut1 - cut2).array()
+                               + m_log_1p_exp_cut1))
+               .sum();
+  } else {
+    if (y_seq[0] == 1) {
+      logp = m_log_1p_exp_cut1.sum();
+    } else if (y_seq[0] == N_classes) {
+      logp = m_log_1p_exp_m_cut2.sum();
+    } else {
+      logp = (m_log_1p_exp_m_cut2 + log1m_exp(cut1 - cut2).array()
+              + m_log_1p_exp_cut1)
+                 .sum();
+    }
+  }
+
+  operands_and_partials<Matrix<T_x_scalar, T_x_rows, Dynamic>,
+                        Eigen::Matrix<T_beta_scalar, Eigen::Dynamic, 1>,
+                        Eigen::Matrix<T_cuts_scalar, Eigen::Dynamic, 1>>
+      ops_partials(x, beta, cuts);
+  if (!is_constant_all<T_x_scalar, T_beta_scalar, T_cuts_scalar>::value) {
```

Comment:
I don't think you want ```!is_constant_all<>``` here, since you'll want to calculate gradients in the situation where some (but not all) are autodiff types

I wonder if an ```is_constant_any<>``` could be a good idea for the future

Reply:
That if says "if not all are constant", which is equivalent to "is any non-constant". This is exactely as intended, as we need derivatives only for non-constant variables.

Reply:
Right you are